File: src/pup/tracingStructs.ml
open PupCore.PupTypes
open PupCore.Pup

let ticktock = structure [("start",  U64);
                          ("end",    U64);
                          ("tag_sz", U64);
                          ("tag",    runtime_array I8 "tag_sz" 64);]  (* null-terminated string *)

let next_image = structure [("next_image_idx", U32);]

let vk_trace = structure [("stage_sz", U32);
                          ("stage",    runtime_array I8 "stage_sz" 16); (* if the size isn't enforced anywhere, weird things happen *)
                          ("dt",       U64);]

let doc = document [("ticktock", ticktock);
                    ("next_image", next_image);
                    ("vk_trace", vk_trace);]

(* not supported in python but seems to be working in C *)
let test = structure [("a_sz", U64);
                      ("cnt", U64);
                      ("arr", Array
                         (Array (U64,
                                 (RuntimeSize (Member "a_sz", Constant 10))),
                          (RuntimeSize (Member "cnt", Constant 10))));]

File: src/tracing_structs.py
import struct
class Ticktock:
  __slots__ = ['start', 'end', 'tag_sz', 'tag']
  @classmethod
  def unpack(cls, b):
    (start,) = struct.unpack('=Q', b[(0):((0) + (8))])
    (end,) = struct.unpack('=Q', b[((0) + (8)):(((0) + (8)) + (8))])
    (tag_sz,) = struct.unpack('=Q', b[(((0) + (8)) + (8)):((((0) + (8)) + (8)) + (8))])
    tag = []
    for i in range(0, tag_sz):
      (tmp,) = struct.unpack('=b', b[(((((0) + (8)) + (8)) + (8)) + ((1) * (i))):((((((0) + (8)) + (8)) + (8)) + ((1) * (i))) + (1))])
      tag.append(tmp)
    return cls(start, end, tag_sz, tag)


  def __init__(self, start, end, tag_sz, tag):
    self.start = start
    self.end = end
    self.tag_sz = tag_sz
    self.tag = tag

  def pack(self):
    pass

  def __str__(self):
   return f'Ticktock(start={self.start},end={self.end},tag_sz={self.tag_sz},tag={self.tag})'

class NextImage:
  __slots__ = ['next_image_idx']
  @classmethod
  def unpack(cls, b):
    (next_image_idx,) = struct.unpack('=I', b[(0):((0) + (4))])
    return cls(next_image_idx)


  def __init__(self, next_image_idx):
    self.next_image_idx = next_image_idx

  def pack(self):
    pass

  def __str__(self):
   return f'NextImage(next_image_idx={self.next_image_idx})'

class VkTrace:
  __slots__ = ['stage_sz', 'stage', 'dt']
  @classmethod
  def unpack(cls, b):
    (stage_sz,) = struct.unpack('=I', b[(0):((0) + (4))])
    stage = []
    for i in range(0, stage_sz):
      (tmp,) = struct.unpack('=b', b[(((0) + (4)) + ((1) * (i))):((((0) + (4)) + ((1) * (i))) + (1))])
      stage.append(tmp)
    (dt,) = struct.unpack('=I', b[(((0) + (4)) + ((1) * (stage_sz))):((((0) + (4)) + ((1) * (stage_sz))) + (4))])
    return cls(stage_sz, stage, dt)


  def __init__(self, stage_sz, stage, dt):
    self.stage_sz = stage_sz
    self.stage = stage
    self.dt = dt

  def pack(self):
    pass

  def __str__(self):
   return f'VkTrace(stage_sz={self.stage_sz},stage={self.stage},dt={self.dt})'

PUP_STRUCT_IDS = {
  0: Ticktock,
  1: NextImage,
  2: VkTrace
}